### ISSUE TITLE
OneTrainer - TheRock Setup - Train Device Set to CPU / No Feedback PyTorch in Use (w11, 9070xt)

### ISSUE THREAD

**Issue by xenlaRZ** [CS1o/Stable-Diffusion-Info#57]
Hi, I've been following your [guide](https://github.com/CS1o/Stable-Diffusion-Info/wiki/Lora-Trainer-Setup-Guides#amd-onetrainer-with-rocmtherock) for running OneTrainer with The Rock.

Depending on what option is selected when installing OneTrainer (AMD GPU on Windows query):
- Selecting "yes" for AMD on Windows option - OneTrainer will attempt to use ZLUDA implementation (which tends to fail, unless I try lshqqytiger's fork, however I've found that to run very slow compared to what my 6800XT was achieving on a much older version of OneTrainer)
- Selecting 'no' for AMD on Windows option - OneTrainer will return an error "Found no NVIDIA driver on your system"

For the latter option, I noticed a prior warning also:
"E:\AI\Local\OneTrainer\venv\Lib\site-packages\torch\amp\grad_scaler.py:136: UserWarning: torch.cuda.amp.GradScaler is enabled, but CUDA is not available.  Disabling."

For the 9070XT, I've used [https://github.com/scottt/rocm-TheRock/releases/tag/v6.5.0rc-pytorch-gfx110x](this) set of Pyrotch wheels (Python3.12 files). exporting a debug log from OneTrainer suggests torch==2.7.1+cu128 is installed, however if i try to uninstall torch manually (per step 10 of the guide) it seems to correctly identify the version from the custom Pytorch wheel.

Appreciate any information to get this going, especially:
- Confirming that during the OneTrainer installation I should be using the 'yes' response to the AMD GPU on Windows query.
- Any way to confirm/debug that the command to enable Triton in the start-ui.bat is taking effect

Many thanks in advance.

**Comment by xenlaRZ** [comment]
Getting closer, and certainly no errors whilst using OneTrainer but we may have lobotomised the software!

- On install of OneTrainer, responded 'n' for final query "AMD GPU?" The installation process will initial report this as an error but will complete.
- Followed the steps per the guide, noting the reverse order of custom torch wheel installation. Just as a precaution, repeated numpy installation step with --force-reinstall
- Ran `pip uninstall bitsandbytes`
- OneTrainer will start without errors. Manually set Train Device as 'cuda' under the general tab (whether this makes a difference I'm not sure, but the only alternative is 'cpu' which would ensure the GPU is not used).
- Running a test to train an SDXL lora, no errors reported, fast iterations (in relative terms, ~1.5-1.9 it/s vs ~9s/it with 6800XT using zluda).
- Note -  Whilst running, loss rates are not being correctly reported (loss=0, smooth loss=0). This is suggesting that no training is actually being done, and would explain the iteration speed
- Note - AMD Adrenalin driver (v25.10.2) reports a driver timeout when the first iteration starts. The system recovers in my case at least. The behaviour is consistent across all my test runs, but the actual training continues.
- Testing the lora in sdnext (confirmed functioning with older lora's) the output images are not affected by the test run loras.

It might get tricky from this point to understand what is falling over. Happy to look at any suggestions!

**Comment by xenlaRZ** [comment]
Just an exercise in curiosity, I tried using the pytorch builds from the official windows preview following [this](https://gpuopen.com/learn/pytorch-windows-amd-llm-guide/) guide. Namely, installing the following into the OneTrainer venv:

```
pip install --no-cache-dir https://repo.radeon.com/rocm/windows/rocm-rel-6.4.4/torch-2.8.0a0%2Bgitfc14c65-cp312-cp312-win_amd64.whl
pip install --no-cache-dir https://repo.radeon.com/rocm/windows/rocm-rel-6.4.4/torchaudio-2.6.0a0%2B1a8f621-cp312-cp312-win_amd64.whl
pip install --no-cache-dir https://repo.radeon.com/rocm/windows/rocm-rel-6.4.4/torchvision-0.24.0a0%2Bc85f008-cp312-cp312-win_amd64.whl
pip install transformers accelerate
```

Installing over the top of the previous files, most dependencies just show up as already having been met. As expected on starting training (based on the AMD guide above at least), the following error is returned:
```
UserWarning: 1Torch was not compiled with memory efficient attention. (Triggered internally at C:\develop\pytorch-test\aten\src\ATen\native\transformers\hip\sdp_utils.cpp:726.)
  hidden_states = F.scaled_dot_product_attention(
```

It does however then proceed to use the GPU for training without the driver timing out, albeit extremely slowly (200-400 s/it). In my case at least, it makes me think something from the torch wheel proposed in the guide conflicting with the GPU drivers is the culprit but I'm not sure how to diagnose that further.

**Comment by CS1o** [comment]
Hey sry for the late response.
Haven't had the time to look into OneTrainer currently.
The error seems that torch doesnt work with SDP Attention enabled.
Also this strange path here `C:\develop\pytorch-test\aten\src\ATen\native\transformers\hip\sdp_utils.cpp`, does it exists on your system?

I will test around a bit the next days probably. Will report back.

**Comment by xenlaRZ** [comment]
Hi again, had another go recently, and things are working (aside from one driver timeout). As above, rather than using the wheels in the guide under step 10, I followed steps 1 and 2 from the AMD guide below:
[PyTorch via PIP installation](https://rocm.docs.amd.com/projects/radeon-ryzen/en/latest/docs/install/installrad/windows/install-pytorch.html)
This installs the ROCm environment and PyTorch from AMD themselves. Whilst still in the OneTrainer virtual environment, the subsequent steps from the guide above can be used to confirm the installation as well as correct detection of the GPU.

When starting OneTrainer there are still a couple of errors (such as bits and bytes discussed before) however these do not prevent training to occur. At least in my testing with SDXL loras. With an appropriate lora rank and alpha the training time is reasonable at 3-5 seconds per iteration..

Note this will only work with the latest AMD driver (26.1.1). I think in my case this resolves the issue, Closing out!

**Comment by CS1o** [comment]
Hey, thanks for letting me know. AMD newly added the RDNA 2 to their native supported gpus for ROCm.
So i will update the guides in the next days probably. That will include installing the latest supported pytorch+rocm and bitsandbytes and solves most issues then.

**Comment by xenlaRZ** [comment]
Hi @CS1o , sorry to bump a closed issue! Noticed the updated instructions for OneTrainer. Have tried and believe that all works as intended (including bits and bytes). Worth noting there is an issue in the recent build of ROCm and the MIOpen library for consumer GPUs like the 9070xt.

At some point in training I see the following error being reported (hundreds of times):
[MIOpen] "Unable to load file: gfx908_metadata.tn.model"

The constant logging in OneTrainer tends to interrupt the actual training process in my case. It's at least an understood problem:
[[Issue]: [MIOpen] "Unable to load file: gfx908_metadata.tn.model" (on consumer grade hardware) #4790](https://github.com/ROCm/rocm-libraries/issues/4790)
I see the same in other applications like sdnext.

Given it's not a OneTrainer issue, might be a case of needing a fix in MIOpen before it goes away.

**Comment by CS1o** [comment]
Hey @xenlaRZ , thanks for letting me know.
Hopefully AMD fixes that soon. Meanwhile can you test if the logs go away when you add this to the start-ui.bat:
`set MIOPEN_ENABLE_LOGGING=0`
`set MIOPEN_ENABLE_LOGGING_CMD=0`
`set MIOPEN_LOG_LEVEL=1`

**Comment by xenlaRZ** [comment]
Confirmed! Including those commands eliminates the logging issue. Prior to using them I was getting around 5 iterations into training and then having all progress stall as the logging became continuous. Have managed to train ~3000 iterations without issue since adding those. Thank you!
